Remove reference to `defaultcomponents` in core and deprecate `include_core` flag

JamesJHPark opened the pull request:
Description:
This PR is to remove reference to `defaultcomponents` in core and deprecate the `include_core` flag in builder.

Link to tracking Issue:
Issue #3927

dashpole reviewed (commented):
This removes a lot of unit tests.  Is there a reason they are no longer needed, or should we add them back in in the new location?

bogdandrutu commented:
@dashpole they are already copied to contrib https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/components/extensions_test.go

dashpole reviewed (approved):
Got it, thanks.

bogdandrutu commented on .github/workflows/build-and-test.yml:
```diff
@@ -81,7 +81,7 @@ jobs:
           path: /home/runner/go/bin
           key: tools-${{ runner.os }}-${{ hashFiles('./internal/tools/go.mod', './cmd/mdatagen/go.mod', './cmd/mdatagen/*.go') }}
       - name: Lint
-        run: make -j5 checklicense misspell checkdoc goimpi golint
```
Found it in the contrib.

JamesJHPark replied:
Thank you for pointing it out, I have added `checkdoc` to the test to keep.

bogdandrutu commented on cmd/otelcol/main.go:
```diff
@@ -21,13 +21,13 @@ import (
 
 func main() {
-	factories, err := defaultcomponents.Components()
+	factories, err := testcomponents.DefaultComponents()
```
Cannot depend on a "testing" package from the main file.

Aneurysm9 replied:
Isn't this going away anyways?  Does it matter?

bogdandrutu commented on internal/testcomponents/example_factories.go:
```diff
@@ -39,3 +48,52 @@ func ExampleComponents() (
 }
+
+// DefaultComponents returns the default set of components used by the OpenTelemetry collector.
+func DefaultComponents() (
```
probably this needs to be in a different package than the example/testing components.

Aneurysm9 replied:
Does this need to live elsewhere for `cmd/otelcol` to access it?  That is going away.  If it needs to live elsewhere for other things to access then I think we need a shared understanding of what those things are and how they will utilize it.

JamesJHPark replied:
@bogdandrutu This has been moved to internal package. Thank you.

anuraaga commented:
Talked with @JamesJHPark offline a little to understand this PR - to be honest I'm just jumping in so may be missing something. But the recommendation that made sense to me is

1) Move defaultcomponents into `cmd/otelcol` - it seems to be the configuration for that command specifically and could be local to it. And when the col gets deleted defaultcomponents goes with it

2) Don't reference defaultcomponents from other tests. For example this one

https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/collector_test.go#L75

There doesn't seem to be any reason to couple this to default components. The test could probably just define a couple of components inline. Alternatively, a simple `testcomponents` could also be added, which isn't defaultcomponents (which is moved to otelcol) but something similar, probably smaller, only for tests. I don't know if there would be big value in such a package though vs just inlining the definition of a component or two in the tests that need it.

bogdandrutu commented:
@anuraaga completely agree with your comment. We should not depend on the real components in tests as you suggested.

JamesJHPark commented:
Hi @bogdandrutu, with reference to the discussions, I have made the following changes: 

1. Partially removed dependency on the real components and removed reference to `defaultcomponents` in tests (exception: `cmd/otelcol/main.go` file). For discussion, the function call to `assertZPages` in the [test](https://github.com/open-telemetry/opentelemetry-collector/blob/ba8bc42a61d60e68912947c3c898ce85d7477ce0/service/collector_test.go#L144) requires zpages extension. To address this, I have extracted the components of [testdata/otel-config.yaml](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/testdata/otelcol-config.yaml) to a function in `testcomponents` package for this particular test. This function is also called in `collector_windows_test.go` file. 
2. Since the collector-builder and collector-contrib depend on `defaultcomponents` from the core repository, I have not removed the `defaultcomponents` package. I have removed the tests in this package as they have already been copied to contrib.  

Please let me know if there may be a recommended approach/fixes to address the issue for next steps. Thank you.

alolita commented:
@Aneurysm9 @codeboten can you pl re-review?

Aneurysm9 commented on CHANGELOG.md:
```diff
@@ -30,6 +30,7 @@
 - otlpexporter: Do not retry on PermissionDenied and Unauthenticated (#4349)
 - Remove deprecated funcs `consumererror.As[Traces|Metrics|Logs]` (#4364)
 - Remove support to expand env variables in default configs (#4366)
+- Remove `defaultcomponents` from core (#4087)
```
This needs to be moved up to the `Unreleased` section.

jpkrohling commented:
As discussed during the SIG meeting today, we need first to get the builder to deprecate/remove the `include_core` flag, as new distributions will then not have core components in by default. When that flag is set to `true`, we use the default components, hence the dependency. Contrib might also use the default components, so, that needs changing as well. Once we have a release of the builder out with the new default, we can then merge this.

JamesJHPark commented on CHANGELOG.md:
```diff
@@ -4,6 +4,12 @@
 
+## 🧰 Bug fixes 🧰
+
+- Fix handling of corrupted records by persistent buffer (experimental) (#4475)
```
Thank you for pointing this out. I will file the changes relevant to the PR in a separate commit.

Aneurysm9 commented on cmd/builder/internal/command.go:
```diff
@@ -61,7 +61,8 @@ func Command() (*cobra.Command, error) {
 	cmd.Flags().StringVar(&cfg.Distribution.Version, "version", "1.0.0", "The version for the OpenTelemetry Collector distribution")
-	cmd.Flags().BoolVar(&cfg.Distribution.IncludeCore, "include-core", true, "Whether the core components should be included in the distribution")
+	// IncludeCore is deprecated and will be removed in a subsequent release
+	cmd.Flags().BoolVar(cfg.Distribution.IncludeCore, "include-core", true, "Whether the core components should be included in the distribution")
```
```suggestion
	cmd.Flags().BoolVar(cfg.Distribution.IncludeCore, "include-core", true, "Deprecated: Whether the core components should be included in the distribution")
```

JamesJHPark replied:
Thank you, this has been fixed.

JamesJHPark commented on cmd/builder/README.md:
```diff
@@ -73,7 +73,6 @@ dist:
     module: github.com/open-telemetry/opentelemetry-collector # the module name for the new distribution, following Go mod conventions. Optional, but recommended.
     name: otelcol-custom # the binary name. Optional.
     description: "Custom OpenTelemetry Collector distribution" # a long name for the application. Optional.
-    include_core: true # whether the core components should be included in the distribution. Optional.
```
Yes, I have implemented the change to deprecate this before removing it.

JamesJHPark replied:
This property has been removed from the example.

JamesJHPark replied:
This property has been removed from the example.

JamesJHPark commented on cmd/builder/internal/builder/config.go:
```diff
@@ -54,7 +54,6 @@ type Distribution struct {
 	Go             string `mapstructure:"go"`
 	Description    string `mapstructure:"description"`
 	OtelColVersion string `mapstructure:"otelcol_version"`
-	IncludeCore    bool   `mapstructure:"include_core"`
```
Hi, I have set `IncludeCore` to `*bool` with a comment to state that this will be removed in a subsequent release. However, setting this field to `*bool` caused the nil pointer dereference issue, specifically in a function `newBoolValue` from `BoolVar` in this [file](https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/builder/internal/command.go#L64). I have tested to set `IncludeCore` to an address of temporary (for deprecation) variable for `nil`, but the test for `nocore.otel.yaml` failed and with duplicate components. Please let me know if there may be an approach I can take to ensure the Builder Integration tests succeed in this case.

JamesJHPark commented:
> You might need to set the default value of the flag in `DefaultConfig`. This would solve your nil pointer error. Also, make sure the distribution is indeed _not_ having the default components, as I believe there might be a change needed in the templates.

Thank you for your reply. I have set the default value of the flag in the `DefaultConfig` to fix the nil pointer dereference issue. Upon investigating further, the nocore test failed with duplicate components. In `nocore.builder.yaml`, `include_core` is set to `false`, where this flag has been changed to `*bool`.  

I have tested and set `*cfg.Distribution.IncludeCore = false`, specifically for the failed nocore test, after unmarshalling config in this [file](https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/builder/internal/command.go#L106-L109). However, the nocore test did not pass with duplicate components, and this may be related to the issue of dereferencing `.Distribution.IncludeCore` to a boolean value (also tested to use this output value in the pipeline for eq comparison) in the go.tmpl [file](https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/builder/internal/builder/templates/components.go.tmpl#L38). Since the flag was set to `*bool`, I have also tried and set `cfg.Distribution.IncludeCore` to `nil`. Although the tests (replaces, nocore) passed, this would indicate that there is a tri-state (true, false, nil) for the flag but its state would be determined by `nil` and `non-nil` values. Therefore, another suggested approach that I have currently implemented is to include the warn log before unmarshalling the config in this [file](https://github.com/open-telemetry/opentelemetry-collector/blob/4433904d165bb249299a05340ba03cb3355dd3dc/cmd/builder/internal/command.go#L110-L113) and the `MarkDeprecated` flag message in this [file](https://github.com/open-telemetry/opentelemetry-collector/blob/4433904d165bb249299a05340ba03cb3355dd3dc/cmd/builder/internal/command.go#L66-L68) with `IncludeCore` set back to `bool`. Additionally, I have not modified the default components in `components.go.tmpl` for the `include_core` deprecation. Please let me know if these changes may be suitable to achieve the same goals to deprecate the `include_core` flag.

JamesJHPark commented on cmd/builder/internal/command.go:
```diff
@@ -61,7 +61,8 @@ func Command() (*cobra.Command, error) {
 	cmd.Flags().StringVar(&cfg.Distribution.Version, "version", "1.0.0", "The version for the OpenTelemetry Collector distribution")
-	cmd.Flags().BoolVar(&cfg.Distribution.IncludeCore, "include-core", true, "Whether the core components should be included in the distribution")
+	// IncludeCore is deprecated and will be removed in a subsequent release
+	cmd.Flags().BoolVar(cfg.Distribution.IncludeCore, "include-core", true, "Deprecated: Whether the core components should be included in the distribution")
```
This has been fixed.

jpkrohling replied:
Sorry, my bad, this should be 0.41.0 instead of 1.41.0.

JamesJHPark replied:
Thank you for clarification.

jpkrohling commented on cmd/builder/internal/command.go:
```diff
@@ -102,6 +107,11 @@ func initConfig() error {
 
+	// if the include_core flag is set from the config, create a warning message that this is deprecated and will be removed in a subsequent release
+	if vp.IsSet("dist.include_core") {
+		cfg.Logger.Warn("IncludeCore is deprecated, starting from v1.41.0, core components are not going to be implicitly included ")
```
On a second thought, I think we need to have this message even if people are not using this flag, as not having the flag implies that core is included, and we are changing the behavior. Perhaps we can have something like the following as part of `config#Validate`.

```go
	// create a warning message that include_core is deprecated and will be removed in a subsequent release
	if c.Distribution.IncludeCore {
		c.Logger.Warn("IncludeCore is deprecated. Starting from v0.41.0, you need to include all components explicitly.")
	} else {
		c.Logger.Info("IncludeCore is deprecated, starting from v0.41.0. The new behavior won't affect your distribution, just remove the option from the manifest.")
	}
```

JamesJHPark replied:
Yes, thank you for pointing this out. I have included this as part of `config#Validate`.

jpkrohling commented on internal/testcomponents/example_factories.go:
```diff
@@ -39,3 +45,40 @@ func ExampleComponents() (
 }
+
+// OtelColConfigComponents returns the set of components in "testdata/otelcol-config.yaml". This is only used by tests.
+func OtelColConfigComponents() (
```
This function name can probably be changed to something like `DefaultFactories`, which becomes this to callers: `testcomponents.DefaultFactories`.

JamesJHPark replied:
I have changed the function name to `DefaultFactories`.

jpkrohling reviewed (commented):
I think those are my final concerns, after this, this will be ready to be merged.

sodabrew commented:
I no longer understand what to put in my builder.yml to get the core modules back, or which ones were default-in that I was relying upon without knowing it. Would appreciate more clarity in the builder's README file on this.

jpkrohling commented:
I apologize, this change should have been better communicated. This PR brought clarify to the readme: https://github.com/open-telemetry/opentelemetry-collector/pull/4664

If you think there's still something we should document elsewhere, let me know.

sodabrew commented:
Thank you for the README example improvement!
